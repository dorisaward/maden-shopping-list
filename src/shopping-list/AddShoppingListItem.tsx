import React, {useCallback, useState} from 'react';
import {GroceryItem} from '../groceries/GroceryItem';
import {View, TextInput, TouchableOpacity, Text, Alert} from 'react-native';
import {styles} from '../styles';
import uuid from 'react-native-uuid';

type Props = {
  addGroceryItem: (groceryItem: GroceryItem) => void;
};

export const AddShoppingListItem = ({
  addGroceryItem,
}: Props): React.JSX.Element => {
  const [newGroceryName, setNewGroceryName] = useState<string | undefined>(
    undefined,
  );
  const [newGroceryQuantity, setNewGroceryQuantity] = useState<number>(0);

  const setQuantity = useCallback(
    (text: string) => {
      try {
        const number = Number(text);
        setNewGroceryQuantity(number);
      } catch (e) {
        console.warn(`Could not cast string ${text} to number`, e);
      }
    },
    [setNewGroceryQuantity],
  );

  const saveGroceryItem = useCallback(() => {
    if (!newGroceryName) {
      Alert.alert('Please make sure your groceries have a name');
      return;
    }
    if (!newGroceryQuantity) {
      Alert.alert(`Please specify how much ${newGroceryName} you want.`);
      return;
    }
    addGroceryItem({
      name: newGroceryName,
      quantity: newGroceryQuantity,
      id: uuid.v4().toString(),
    });
    setNewGroceryQuantity(0);
    setNewGroceryName(undefined);
  }, [newGroceryName, newGroceryQuantity, addGroceryItem]);

  return (
    <View style={styles.shoppingListItem}>
      <TextInput
        placeholder="Name"
        value={newGroceryName}
        style={styles.shoppingListItemInput}
        onChangeText={setNewGroceryName}
      />
      <TextInput
        placeholder="Quantity"
        value={newGroceryQuantity.toString()}
        keyboardType="number-pad"
        style={styles.shoppingListItemInput}
        onChangeText={setQuantity}
      />
      <TouchableOpacity style={styles.saveText} onPress={saveGroceryItem}>
        <Text>Save</Text>
      </TouchableOpacity>
    </View>
  );
};
