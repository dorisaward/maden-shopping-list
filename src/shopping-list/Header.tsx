import {Text, View} from 'react-native';
import {styles} from '../styles';
import React from 'react';

export const Header = (): React.JSX.Element => (
  <View style={styles.background}>
    <Text style={styles.headerText}>Mayden Shopping List</Text>
  </View>
);
