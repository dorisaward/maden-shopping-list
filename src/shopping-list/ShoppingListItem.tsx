import React from 'react';
import {ListRenderItem, Text, View} from 'react-native';
import {GroceryItem} from '../groceries/GroceryItem';
import {styles} from '../styles';

export const ShoppingListItem: ListRenderItem<GroceryItem> = ({item}) => (
  <View style={styles.shoppingListItem}>
    <Text style={styles.shoppingListItemText}>{item.name}</Text>
    <Text style={styles.shoppingListItemText}>{item.quantity}</Text>
  </View>
);
