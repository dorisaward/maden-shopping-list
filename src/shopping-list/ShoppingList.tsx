import React, { useState } from 'react';
import {FlatList, FlatListProps} from 'react-native';
import {ShoppingListItem} from './ShoppingListItem';
import {GroceryItem} from '../groceries/GroceryItem';

type Props = {
  initialGroceryItems: GroceryItem[];
};

export const ShoppingList = ({
  initialGroceryItems,
}: Props): React.JSX.Element => {
const [groceryItems, setGroceryItems] = useState<GroceryItem[]>(initialGroceryItems)

  return (
    <FlatList
      data={groceryItems}
      renderItem={ShoppingListItem}
      keyExtractor={keyExtractor}
    />
  );
};

const keyExtractor: FlatListProps<GroceryItem>['keyExtractor'] = item =>
  item.id;
