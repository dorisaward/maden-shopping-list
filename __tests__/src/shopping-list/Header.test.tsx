import 'react-native';
import React from 'react';
import {it, expect} from '@jest/globals';
import {render} from '@testing-library/react-native';
import {Header} from '../../../src/shopping-list/header';

it('should render header', () => {
  // Given
  const renderable = <Header />;

  // When
  const {toJSON} = render(renderable);

  // Then
  expect(toJSON()).toMatchSnapshot();
});
