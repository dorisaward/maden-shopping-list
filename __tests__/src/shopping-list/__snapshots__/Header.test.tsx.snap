// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render header 1`] = `
<View
  style={
    {
      "backgroundColor": "cream",
      "padding": 10,
    }
  }
>
  <Text
    style={
      {
        "color": "black",
        "fontSize": 30,
        "textAlign": "center",
      }
    }
  >
    Mayden Shopping List
  </Text>
</View>
`;
